File: src/presets.ts
import {validateChunk, validateSearch, type ChunkConfig, type SearchConfig, type MnemeConfig} from "./defaults.js"

export interface Preset
{
    chunk: ChunkConfig
    search: SearchConfig
}

interface SweepAxes
{
    chunkSize: number[]
    overlap: number[]
    alpha: number[]
    k: number[]
}

export function getPreset(name: string, base: MnemeConfig): Preset[]
{
    if (name === "fast") return expand({
        chunkSize: [base.chunk.chunkSize],
        overlap:   [base.chunk.overlap],
        alpha:     [base.search.alpha],
        k:         [base.search.k],
    })
    if (name === "medium") return expand({
        chunkSize: [base.chunk.chunkSize],
        overlap:   [base.chunk.overlap],
        alpha:     [0.0, 0.3, 0.5, 0.7, 1.0],
        k:         [5, 10, 20],
    })
    if (name === "thorough") return expand({
        chunkSize: [500, 1000],
        overlap:   [0, 0.2],
        alpha:     [0.5, 0.7],
        k:         [5],
    })
    throw new Error(`unknown sweep level: ${name} (expected: fast | medium | thorough)`)
}

function expand(axes: SweepAxes): Preset[]
{
    const out: Preset[] = []
    for (const chunkSize of axes.chunkSize)
        for (const overlap of axes.overlap)
            for (const alpha of axes.alpha)
                for (const k of axes.k)
                {
                    const p: Preset = {chunk: {chunkSize, overlap}, search: {alpha, k}}
                    validateChunk(p.chunk)
                    validateSearch(p.search)
                    out.push(p)
                }
    return out
}

File: mneme/eval/presets.py
from dataclasses import replace
from itertools import product

from ..core.config import MnemeConfig, resolve_config


def get_preset(level: str, base: MnemeConfig) -> list[MnemeConfig]:
    """Expands a sweep level into a list of concrete configs to evaluate.
    Each returned config is a copy of `base` with one combination of
    chunk_size/overlap/alpha/k applied."""
    if level == "fast":
        return _expand(base, [base.chunk_size], [base.overlap], [base.alpha], [base.k])
    if level == "medium":
        return _expand(base, [base.chunk_size], [base.overlap], [0.0, 0.3, 0.5, 0.7, 1.0], [5, 10, 20])
    if level == "thorough":
        return _expand(base, [500, 1000], [0.0, 0.2], [0.0, 0.3, 0.5, 0.7, 1.0], [5, 10, 20])
    raise ValueError(f"unknown sweep level: {level} (expected: fast | medium | thorough)")


def _expand(base: MnemeConfig, chunk_sizes, overlaps, alphas, ks) -> list[MnemeConfig]:
    out: list[MnemeConfig] = []
    for chunk_size, overlap, alpha, k in product(chunk_sizes, overlaps, alphas, ks):
        cfg = replace(base, chunk_size=chunk_size, overlap=overlap, alpha=alpha, k=k)
        resolve_config(cfg)
        out.append(cfg)
    return out
